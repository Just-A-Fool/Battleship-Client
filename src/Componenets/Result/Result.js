import React, { Component } from 'react';
import { Link } from 'react-router-dom';
import './Result.css';

export default class Result extends Component {

  render() {
    return (

    );
  };
  
};


