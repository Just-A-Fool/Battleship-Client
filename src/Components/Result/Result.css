.result {
    margin-top: 8rem;
}

.results-ul {
    border: 0.1rem solid white; 
    width: 65%;
    margin: 0 auto;
    padding: 10px;
    border-radius: 1rem;
    background-color: rgb(30, 52, 90);
}

.results-ul li {
    display: flex;
    margin: 5px 0;
}

.results-left {
    width: 50%;
    text-align: right;
    padding-right: 5px;
}

.results-right {
    width: 50%;
    text-align: left;
    padding-left: 5px;
}