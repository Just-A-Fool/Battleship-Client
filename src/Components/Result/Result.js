import React, { Component } from 'react';
import { withRouter } from 'react-router-dom';
import LoadGameApiService from '../../Services/load-game-api-service';
import Button from '../Button/Button';
import Header from '../Header/Header';
import './Result.css';


class Result extends Component {

  static defaultProps = {
    player: null,
    game: null
  }

  constructor(props) {
    super(props);
    this.state={
      error:null,
      player:'',              // player1 or player2
      game:'',                // game ID
      winner:'',              // Winner of game, as player1 or player2
      loser:'',               // Loser of game (if not a tie), as player1 or player2
      hRatio:'',              // Player hits divided by total shots taken
      totalShots:'',          // Total shots fired at opponent (hits + misses)
      totalMisses:'',         // Total shots at opponent (misses)
      totalHits:'',           // Total shots at opponent (hits)
      player1_hits:[],        // Array holding values of first player hits (alphanumeric cell values)
      player1_misses:[],      // Array holding values of first player misses       ''      ''   
      player2_hits:[],        // Array holding values of second player hits (alphanumeric cell values)
      player2_misses:[],      // Array holding values of second player misses      ''      ''
    }
  }

  /*
      If an error occurs while component is loading, we are setting it here in state
  */
  setError = (err) => {
    this.setState({
      error: err.error
    })
  }

  componentDidMount() {
    if (this.props.results) {
      const player = this.props.results.player;
      const game = this.props.results.game;
      LoadGameApiService.getResults(game)
        .then((res) => {

          let p1Hits = JSON.parse(res[0].player1_hits);
          let p1Misses = JSON.parse(res[0].player1_misses);
          let p2Hits = JSON.parse(res[0].player2_hits);
          let p2Misses = JSON.parse(res[0].player2_misses);

          this.setState({
            player1_hits: p1Hits ? p1Hits: [],
            player1_misses: p1Misses ? p1Misses: [],
            player2_hits: p2Hits ? p2Hits: [],
            player2_misses: p2Misses ? p2Misses: [],
            player: player
          },() => this.updateResults())
          
        })
        .catch((e) => this.setError(e));
    }
  }

  updateResults = () => {
    let playerString = this.props.results.player;


    let hits = this.state[`${playerString}_hits`].length;
    let misses = this.state[`${playerString}_misses`].length;
    let totalShots = hits + misses;
    let hitRatio = hits / totalShots;
    let winnerUsername = this.props.results.playerWin ? this.props.results.playerUsername : this.props.results.opponentUsername;
    let loserUsername = this.props.results.playerWin ? this.props.results.opponentUsername : this.props.results.playerUsername;

    this.setState({
      winner: winnerUsername,
      loser: loserUsername,
      hRatio: hitRatio ? hitRatio: 0,
      totalHits: hits,
      totalMisses: misses,
      totalShots: totalShots
    });
  }

  /*
      Render method which displays the results of the previous match. Buttons rendered below results 
      display route to either a new game, a rematch, or back to the dashboard.
  */
  render() {
    return (
      <div className='result'>
        <Header />


        <h1>Result</h1>
        <ul className='results-ul'>

          {this.props.results ? <h2>{`${this.props.results.playerUsername} vs ${this.props.results.opponentUsername}`}</h2> : null}

          <li><span className='results-left'>Winner: </span>
            <span className='results-right'>{this.state.winner}</span></li>

          <li><span className='results-left'>Loser: </span>
            <span className='results-right'>{this.state.loser}</span></li>

          <li><span className='results-left'>Hit Ratio: </span>
            <span className='results-right'>{(this.state.hRatio * 100).toFixed(2)}%</span></li>

          <li><span className='results-left'>Shots Fired: </span>
            <span className='results-right'>{this.state.totalShots}</span></li>

          <li><span className='results-left'>Shots Hit: </span>
            <span className='results-right'>{this.state.totalHits}</span></li>
            
          <li><span className='results-left'>Shots Missed: </span>
            <span className='results-right'>{this.state.totalMisses}</span></li>

        </ul>


        <Button onClick={() => {
          this.props.history.push('/newgame')
        }}> New Game
        </Button>

        <Button onClick={() => {
          this.props.history.push('/dashboard')
        }}> Exit
        </Button>
      </div>
    );
  };
};

export default withRouter(Result);


