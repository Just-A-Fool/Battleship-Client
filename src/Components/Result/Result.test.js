import React from 'react';
import ReactDOM from 'react-dom';
import Result from './Result';
import { mount } from 'enzyme';
import toJson from 'enzyme-to-json';
import { BrowserRouter as Router } from 'react-router-dom';


describe('Help component', () => {
  it('Renders without crashing', () => {
    const div = document.createElement('div');
    ReactDOM.render(<Router><Result /></Router>, div);
    ReactDOM.unmountComponentAtNode(div);
  });

  it(`Displays the 'Result' component when rendered`, () => {
    const wrapper = mount(<Router><Result /></Router>)
    expect(toJson(wrapper)).toMatchSnapshot()
  });

  it(`Contains an 'h1' element for displaying the title`, () => {
    const wrapper = mount(<Router><Result /></Router>)
    expect(wrapper.contains(<h1 className='result-h1'>Result</h1>)).toEqual(true)
  });

  it(`Displays an area for the winner`, () => {
    const wrapper = mount(<Router><Result /></Router>)
    expect(wrapper.text()).toMatch(/Winner:/)
  });

  it(`Displays an area for the loser`, () => {
    const wrapper = mount(<Router><Result /></Router>)
    expect(wrapper.text()).toMatch(/Loser:/)
  });

  it(`Displays an area for the hit ratio`, () => {
    const wrapper = mount(<Router><Result /></Router>)
    expect(wrapper.text()).toMatch(/Hit Ratio:/)
  });

  it(`Displays an area for the total shots fired`, () => {
    const wrapper = mount(<Router><Result /></Router>)
    expect(wrapper.text()).toMatch(/Shots Fired:/)
  });

  it(`Displays an area for the total hits on the enemy vessel`, () => {
    const wrapper = mount(<Router><Result /></Router>)
    expect(wrapper.text()).toMatch(/Shots Hit:/)
  });

  it(`Displays an area for the total misses on the enemy vessel`, () => {
    const wrapper = mount(<Router><Result /></Router>)
    expect(wrapper.text()).toMatch(/Shots Missed:/)
  });

});