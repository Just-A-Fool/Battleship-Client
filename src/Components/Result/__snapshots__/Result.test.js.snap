// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Help component Displays the 'Result' component when rendered 1`] = `
<BrowserRouter>
  <Router
    history={
      Object {
        "action": "POP",
        "block": [Function],
        "createHref": [Function],
        "go": [Function],
        "goBack": [Function],
        "goForward": [Function],
        "length": 1,
        "listen": [Function],
        "location": Object {
          "hash": "",
          "pathname": "/",
          "search": "",
          "state": undefined,
        },
        "push": [Function],
        "replace": [Function],
      }
    }
  >
    <withRouter(Result)>
      <Route>
        <Result
          game={null}
          history={
            Object {
              "action": "POP",
              "block": [Function],
              "createHref": [Function],
              "go": [Function],
              "goBack": [Function],
              "goForward": [Function],
              "length": 1,
              "listen": [Function],
              "location": Object {
                "hash": "",
                "pathname": "/",
                "search": "",
                "state": undefined,
              },
              "push": [Function],
              "replace": [Function],
            }
          }
          location={
            Object {
              "hash": "",
              "pathname": "/",
              "search": "",
              "state": undefined,
            }
          }
          match={
            Object {
              "isExact": true,
              "params": Object {},
              "path": "/",
              "url": "/",
            }
          }
          player={null}
        >
          <div
            className="result"
          >
            <withRouter(Header)>
              <Route>
                <Header
                  history={
                    Object {
                      "action": "POP",
                      "block": [Function],
                      "createHref": [Function],
                      "go": [Function],
                      "goBack": [Function],
                      "goForward": [Function],
                      "length": 1,
                      "listen": [Function],
                      "location": Object {
                        "hash": "",
                        "pathname": "/",
                        "search": "",
                        "state": undefined,
                      },
                      "push": [Function],
                      "replace": [Function],
                    }
                  }
                  location={
                    Object {
                      "hash": "",
                      "pathname": "/",
                      "search": "",
                      "state": undefined,
                    }
                  }
                  match={
                    Object {
                      "isExact": true,
                      "params": Object {},
                      "path": "/",
                      "url": "/",
                    }
                  }
                >
                  <div>
                    <Banner>
                      <div
                        className="banner"
                      >
                        <h1>
                          BATTLESHIP
                        </h1>
                      </div>
                    </Banner>
                    <header
                      className="main-header"
                    >
                      <ForwardRef
                        onClick={[Function]}
                      >
                        <button
                          className="Button"
                          onClick={[Function]}
                        >
                           Dashboard
                        </button>
                      </ForwardRef>
                      <ForwardRef
                        onClick={[Function]}
                      >
                        <button
                          className="Button"
                          onClick={[Function]}
                        >
                           Help
                        </button>
                      </ForwardRef>
                      <ForwardRef
                        onClick={[Function]}
                      >
                        <button
                          className="Button"
                          onClick={[Function]}
                        >
                          Logout
                        </button>
                      </ForwardRef>
                    </header>
                  </div>
                </Header>
              </Route>
            </withRouter(Header)>
            <h1>
              Result
            </h1>
            <ul
              className="results-ul"
            >
              <li>
                <span
                  className="results-left"
                >
                  Winner: 
                </span>
                <span
                  className="results-right"
                />
              </li>
              <li>
                <span
                  className="results-left"
                >
                  Loser: 
                </span>
                <span
                  className="results-right"
                />
              </li>
              <li>
                <span
                  className="results-left"
                >
                  Hit Ratio: 
                </span>
                <span
                  className="results-right"
                >
                  0.00
                  %
                </span>
              </li>
              <li>
                <span
                  className="results-left"
                >
                  Shots Fired: 
                </span>
                <span
                  className="results-right"
                />
              </li>
              <li>
                <span
                  className="results-left"
                >
                  Shots Hit: 
                </span>
                <span
                  className="results-right"
                />
              </li>
              <li>
                <span
                  className="results-left"
                >
                  Shots Missed: 
                </span>
                <span
                  className="results-right"
                />
              </li>
            </ul>
            <ForwardRef
              onClick={[Function]}
            >
              <button
                className="Button"
                onClick={[Function]}
              >
                 Exit
              </button>
            </ForwardRef>
            <Footer>
              <footer>
                Copyright Â© since 2020
              </footer>
            </Footer>
          </div>
        </Result>
      </Route>
    </withRouter(Result)>
  </Router>
</BrowserRouter>
`;
